unit_tests/src/test_client_connection.cpp

```diff
@@ -300,3 +307,108 @@ TEST_CASE("#7192: all connections store unique session signatures", "client_conn
 
     CHECK(std::all_of(std::begin(pairs), std::end(pairs), has_unique_session_signatures));
 }
+
+// NOLINTNEXTLINE(readability-function-cognitive-complexity)
+TEST_CASE("#7155: report underlying error info from server on failed connection")
+{
+    load_client_api_plugins();
+
+    rodsEnv env;
+    _getRodsEnv(env);
+
+    SECTION("invalid host")
+    {
+        try {
+            ix::client_connection{"example.org", env.rodsPort, {env.rodsUserName, env.rodsZone}};
+        }
+        catch (const irods::exception& e) {
+            CHECK(e.code() == USER_SOCK_CONNECT_TIMEDOUT);
+            CHECK_THAT(e.client_display_what(), Catch::Matchers::Contains("_rcConnect: connectToRhost failed"));
+        }
+    }
+
+    SECTION("invalid port")
+    {
+        try {
+            ix::client_connection{env.rodsHost, 8080, {env.rodsUserName, env.rodsZone}};
+        }
+        catch (const irods::exception& e) {
+            // Assertion fails with including the embedded errno value.
+            CHECK(e.code() == (USER_SOCK_CONNECT_ERR - ECONNREFUSED));
```

Comment:
I suppose that this implicitly tests for `ECONNREFUSED` inherently. Pretty slick.

If we wanted to explicitly assert the parts of the error code (i.e. the iRODS error code and the `errno` value), we could use the functions in `irods/system_error.hpp` to extract each bit of information and make assertions. What you have now works, too. Because math.

Reply:
True. Doing the math was easier than pulling in the header lol.

Can change if you feel it's better to use `system_error.hpp`. Either is fine with me.

Just noticed the comment on L336 is weird. Will fix that.

Reply:
Switched to using system_error.hpp.